Integers not parsed as two's complement if not on byte boundary

Integers which don't sit on byte boundaries aren't correctly parsed as two's complement.

If you have a int4 which is "1111" (that's -1 converted), restruct seems to unpack this as "00001111" which is +15 converted.

An example of this in practice can be found here: https://github.com/go-gnss/rtcm/blob/master/rtcm3/antenna.go#L8. The Int64 values in the AntennaReferencePoint type are never correct if the value is supposed to be negative.

If you checkout the code and add the following function to messages_test.go you can see the effect:

```
func TestCoords(t *testing.T) {
    binary := readPayload(1006)
    msg := rtcm3.DeserializeMessage(binary)
    c := msg.(rtcm3.Message1006)
    t.Error("from restruct", c.ReferencePointX, c.ReferencePointY, c.ReferencePointZ)

    x := (int64(binary[4]&63) << 32) | (int64(binary[5]) << 24) | (int64(binary[6]) << 16) | (int64(binary[7]) << 8) | int64(binary[8])
    if binary[4]&32 == 32 {
        x = x - int64(math.Pow(2, 38))
    }
    y := (int64(binary[9]&63) << 32) | (int64(binary[10]) << 24) | (int64(binary[11]) << 16) | (int64(binary[12]) << 8) | int64(binary[13])
    if binary[9]&32 == 32 {
        y = y - int64(math.Pow(2, 38))
    }
    z := (int64(binary[14]&63) << 32) | (int64(binary[15]) << 24) | (int64(binary[16]) << 16) | (int64(binary[17]) << 8) | int64(binary[18])
    if binary[14]&32 == 32 {
        z = z - int64(math.Pow(2, 38))
    }
    t.Error("bit shifting", x, y, z)
}
```

This produces the following result:

```
$ go test .
--- FAIL: TestCoords (0.00s)
    messages_test.go:37: from restruct 230154331576 26704851794 238184162681
    messages_test.go:51: bit shifting -44723575368 26704851794 -36693744263
```

The values from the bit shifting are correct.

--- a/decoder.go
+++ b/decoder.go
@@ -48,10 +48,18 @@ func (d *decoder) readBits(f field, outBuf []byte) {
 		decodedBits = 8 * len(outBuf)
 	} else {
 		decodedBits = int(d.bitSize)
-	}
 
-	// Crop output buffer to relevant bytes only.
-	outBuf = outBuf[len(outBuf)-(decodedBits+7)/8:]
+		// HACK: Go's generic endianness abstraction is not great if you are
+		// working with bits directly. Here we hardcode a case for little endian
+		// because there is no other obvious way to deal with it.
+		//
+		// Crop output buffer to relevant bytes only.
+		if d.order == binary.LittleEndian {
+			outBuf = outBuf[:(decodedBits+7)/8]
+		} else {
+			outBuf = outBuf[len(outBuf)-(decodedBits+7)/8:]
+		}
+	}
 
 	if d.bitCounter == 0 && decodedBits%8 == 0 {
 		// Fast path: we are fully byte-aligned.
@@ -68,37 +76,73 @@ func (d *decoder) readBits(f field, outBuf []byte) {
 	}
 }
 
-func (d *decoder) read8(f field) uint8 {
+func (d *decoder) extend8(val uint8, signed bool) uint8 {
+	if signed && d.bitSize != 0 && val&1<<uint(d.bitSize-1) != 0 {
+		val |= ^((1 << uint(d.bitSize)) - 1)
+	}
+	return val
+}
+
+func (d *decoder) extend16(val uint16, signed bool) uint16 {
+	if signed && d.bitSize != 0 && val&1<<uint(d.bitSize-1) != 0 {
+		val |= ^((1 << uint(d.bitSize)) - 1)
+	}
+	return val
+}
+
+func (d *decoder) extend32(val uint32, signed bool) uint32 {
+	if signed && d.bitSize != 0 && val&1<<uint(d.bitSize-1) != 0 {
+		val |= ^((1 << uint(d.bitSize)) - 1)
+	}
+	return val
+}
+
+func (d *decoder) extend64(val uint64, signed bool) uint64 {
+	if signed && d.bitSize != 0 && val&(1<<uint(d.bitSize-1)) != 0 {
+		val |= ^((1 << uint(d.bitSize)) - 1)
+	}
+	return val
+}
+
+func (d *decoder) read8(f field, extend bool) uint8 {
 	b := make([]byte, 1)
 	d.readBits(f, b)
-	return uint8(b[0])
+	return d.extend8(uint8(b[0]), extend)
 }
 
-func (d *decoder) read16(f field) uint16 {
+func (d *decoder) read16(f field, extend bool) uint16 {
 	b := make([]byte, 2)
 	d.readBits(f, b)
-	return d.order.Uint16(b)
+	return d.extend16(d.order.Uint16(b), extend)
 }
 
-func (d *decoder) read32(f field) uint32 {
+func (d *decoder) read32(f field, extend bool) uint32 {
 	b := make([]byte, 4)
 	d.readBits(f, b)
-	return d.order.Uint32(b)
+	return d.extend32(d.order.Uint32(b), extend)
 }
 
-func (d *decoder) read64(f field) uint64 {
+func (d *decoder) read64(f field, extend bool) uint64 {
 	b := make([]byte, 8)
 	d.readBits(f, b)
-	return d.order.Uint64(b)
+	return d.extend64(d.order.Uint64(b), extend)
 }
 
-func (d *decoder) readS8(f field) int8 { return int8(d.read8(f)) }
+func (d *decoder) readU8(f field) uint8 { return uint8(d.read8(f, false)) }
 
-func (d *decoder) readS16(f field) int16 { return int16(d.read16(f)) }
+func (d *decoder) readU16(f field) uint16 { return uint16(d.read16(f, false)) }
 
-func (d *decoder) readS32(f field) int32 { return int32(d.read32(f)) }
+func (d *decoder) readU32(f field) uint32 { return uint32(d.read32(f, false)) }
 
-func (d *decoder) readS64(f field) int64 { return int64(d.read64(f)) }
+func (d *decoder) readU64(f field) uint64 { return uint64(d.read64(f, false)) }
+
+func (d *decoder) readS8(f field) int8 { return int8(d.read8(f, true)) }
+
+func (d *decoder) readS16(f field) int16 { return int16(d.read16(f, true)) }
+
+func (d *decoder) readS32(f field) int32 { return int32(d.read32(f, true)) }
+
+func (d *decoder) readS64(f field) int64 { return int64(d.read64(f, true)) }
 
 func (d *decoder) readBytes(count int) []byte {
 	x := d.buf[0:count]
@@ -382,28 +426,28 @@ func (d *decoder) read(f field, v reflect.Value) {
 		d.setInt(f, v, d.readS64(f))
 
 	case reflect.Uint8, reflect.Bool:
-		d.setUint(f, v, uint64(d.read8(f)))
+		d.setUint(f, v, uint64(d.readU8(f)))
 	case reflect.Uint16:
-		d.setUint(f, v, uint64(d.read16(f)))
+		d.setUint(f, v, uint64(d.readU16(f)))
 	case reflect.Uint32:
-		d.setUint(f, v, uint64(d.read32(f)))
+		d.setUint(f, v, uint64(d.readU32(f)))
 	case reflect.Uint64:
-		d.setUint(f, v, d.read64(f))
+		d.setUint(f, v, d.readU64(f))
 
 	case reflect.Float32:
-		v.SetFloat(float64(math.Float32frombits(d.read32(f))))
+		v.SetFloat(float64(math.Float32frombits(d.read32(f, false))))
 	case reflect.Float64:
-		v.SetFloat(math.Float64frombits(d.read64(f)))
+		v.SetFloat(math.Float64frombits(d.read64(f, false)))
 
 	case reflect.Complex64:
 		v.SetComplex(complex(
-			float64(math.Float32frombits(d.read32(f))),
-			float64(math.Float32frombits(d.read32(f))),
+			float64(math.Float32frombits(d.read32(f, false))),
+			float64(math.Float32frombits(d.read32(f, false))),
 		))
 	case reflect.Complex128:
 		v.SetComplex(complex(
-			math.Float64frombits(d.read64(f)),
-			math.Float64frombits(d.read64(f)),
+			math.Float64frombits(d.read64(f, false)),
+			math.Float64frombits(d.read64(f, false)),
 		))
 	}
 

--- a/encoder.go
+++ b/encoder.go
@@ -49,10 +49,18 @@ func (e *encoder) writeBits(f field, inBuf []byte) {
 		encodedBits = 8 * len(inBuf)
 	} else {
 		encodedBits = int(e.bitSize)
-	}
 
-	// Crop input buffer to relevant bytes only.
-	inBuf = inBuf[len(inBuf)-(encodedBits+7)/8:]
+		// HACK: Go's generic endianness abstraction is not great if you are
+		// working with bits directly. Here we hardcode a case for little endian
+		// because there is no other obvious way to deal with it.
+		//
+		// Crop input buffer to relevant bytes only.
+		if e.order == binary.LittleEndian {
+			inBuf = inBuf[:(encodedBits+7)/8]
+		} else {
+			inBuf = inBuf[len(inBuf)-(encodedBits+7)/8:]
+		}
+	}
 
 	if e.bitCounter == 0 && encodedBits%8 == 0 {
 		// Fast path: we are fully byte-aligned.

--- a/packing_test.go
+++ b/packing_test.go
@@ -26,7 +26,6 @@ func TestUnpack(t *testing.T) {
 				Dd: 0x12345678,
 			},
 		},
-
 		{
 			data: []byte{
 				0x55, 0x55,
@@ -42,6 +41,114 @@ func TestUnpack(t *testing.T) {
 				B: 0x02,
 				C: 0xAA,
 				D: 0x05,
+			},
+		},
+		{
+			data: []byte{
+				0x55, 0x55,
+			},
+			bitsize: 16,
+			value: struct {
+				A int8 `struct:"int8:3"`
+				B int8 `struct:"int8:2,big"`
+				C int8 `struct:"int8"`
+				D int8 `struct:"int8:3"`
+			}{
+				A: 0x02,
+				B: 0x02,
+				C: -0x56,
+				D: -0x03,
+			},
+		},
+		{
+			data: []byte{
+				0xFF, 0xFF,
+			},
+			bitsize: 16,
+			value: struct {
+				A int8 `struct:"int8:3"`
+				B int8 `struct:"int8:2"`
+				C int8 `struct:"int8"`
+				D int8 `struct:"int8:3"`
+			}{
+				A: -1,
+				B: -1,
+				C: -1,
+				D: -1,
+			},
+		},
+		{
+			data: []byte{
+				0xFF, 0xFF, 0xFF, 0xFF,
+			},
+			bitsize: 32,
+			value: struct {
+				A int32 `struct:"int32:24"`
+				B int8  `struct:"int16:8"`
+			}{A: -1, B: -1},
+		},
+		{
+			data: []byte{
+				0xFF, 0xFF, 0xFF, 0xFF,
+				0xFF, 0xFF, 0xFF, 0xFF,
+			},
+			bitsize: 64,
+			value: struct {
+				A int64 `struct:"int64:49"`
+				B int64 `struct:"int64:15"`
+			}{A: -1, B: -1},
+		},
+		{
+			data: []byte{
+				0x00, 0x00, 0x00, 0x00,
+				0x00, 0x00, 0x80, 0x01,
+			},
+			bitsize: 64,
+			value: struct {
+				A int64 `struct:"int64:49"`
+				B int64 `struct:"int64:15"`
+			}{A: 1, B: 1},
+		},
+		{
+			data: []byte{
+				0x0F, 0x0F, 0x0F, 0x0F,
+				0x0F, 0x0F, 0x0F, 0x0F,
+			},
+			bitsize: 64,
+			value: struct {
+				A int64 `struct:"int64:49"`
+				B int64 `struct:"int64:15"`
+			}{
+				A: 0x00001E1E1E1E1E1E,
+				B: 0x0000000000000F0F,
+			},
+		},
+		{
+			data: []byte{
+				0xAA, 0xBB, 0xCC, 0xDD,
+				0xEE, 0xFF, 0x00, 0x11,
+			},
+			bitsize: 64,
+			value: struct {
+				A uint64 `struct:"uint64:48,little"`
+				B uint64 `struct:"uint16,little"`
+			}{
+				A: 0x0000FFEEDDCCBBAA,
+				B: 0x0000000000001100,
+			},
+		},
+		{
+			data: []byte{
+				0xAA, 0xBB, 0xCC, 0xDD,
+				0xEE, 0xFF, 0x00, 0x11,
+			},
+			bitsize: 64,
+			value: struct {
+				A int64 `struct:"int64:48,little"`
+				B int64 `struct:"int16,little"`
+			}{
+				A: -0x0000001122334456,
+				B: 0x0000000000001100,
 			},
 		},
 		{
@@ -308,7 +415,7 @@ func TestUnpack(t *testing.T) {
 			},
 			bitsize: 64,
 			value: struct {
-				Size  int `struct:"int32,sizeof=Array"`
+				Size  int   `struct:"int32,sizeof=Array"`
 				Array []int `struct:"[]int32"`
 			}{
 				Size:  1,

